### PR TITLE
Adding partial implementation of C++20 concepts and range functions for AZStd::span

### PR DESCRIPTION
**Pull request by lumberyard-employee-dm**
The new concepts to discovered existing issues with the PathIterator and deque::iterator classes
PathIterator wasn't properly an input_iterator and therefore the Path classes weren't a range due to an incorrect const_iterator alias
The deque::iterator classes was missing the operator+ friend function that accepted a (ptrdiff_t, deque::iterator) to fulfill the random_access_iterator concepts

The AZStd implementations of (uninitialized_)copy(_n), (uninitialized_)move(_n) and (uninitialized_)file(_n) have been optimized to use memcpy and memset based on fulfilling the contiguous_iterator concept

Fixed invalid AZStd::vector inserts in FrameGraphExecuter.cpp and SliceditorEntityOwnershipService.cpp
The code was trying to copy the underlying addresses for vector<unique_ptr> to a vector<raw pointer> using insert, which it was doing by using memcpy.

relates #6749

### CODE REVIEW

File: Code/Framework/AzCore/AzCore/std/iterator.h [hunk 1]
```diff
@@ -98,6 +88,8 @@ namespace AZStd
     struct iterator_traits
         : Internal::iterator_traits_type_aliases<Iterator, Internal::has_iterator_type_aliases_v<Iterator>>
     {
+        // Internal type alais meant to indicate that this is the primary template
```
Comment (mbalfour-amzn) [inline Code/Framework/AzCore/AzCore/std/iterator.h:91]: alais -> alias?

File: Code/Framework/AzCore/AzCore/std/ranges/ranges.h [hunk 2]
```diff
@@ -0,0 +1,1157 @@
...
+        constexpr bool has_unqualified_size = false;
+        template <class T>
+        constexpr bool has_unqualified_size<T, void_t<decltype(size(declval<T&>()))>>
+            = !has_member_size<T> && AZStd::Internal::is_class_or_enum<T>;
+
+        template <class T, class = void>
+        constexpr bool has_end_substract_begin = false;
+        template <class T>
+        constexpr bool has_end_substract_begin<T, void_t<decltype(ranges::end(declval<T&>()) - ranges::begin(declval<T&>()))>>
+            = !has_member_size<T> && !has_unqualified_size<T>;
```
Comment (mbalfour-amzn) [inline Code/Framework/AzCore/AzCore/std/ranges/ranges.h:399]: substract -> subtract?

File: Code/Framework/AzCore/AzCore/std/typetraits/common_reference.h [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+namespace AZStd
+{
+    template <class T, class U, template<class> class TQual, template<class> class UQual>
+    struct basic_common_reference
+    {};
+}
+
+namespace AZStd::Internal
+{
+    // const volatile and reference qualifer copy templates
```
Comment (mbalfour-amzn) [inline Code/Framework/AzCore/AzCore/std/typetraits/common_reference.h:33]: qualifer -> qualifier

File: Code/Framework/AzCore/Tests/AZStd/ConceptsTests.cpp [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+        static_assert(!AZStd::totally_ordered_with<Base, TestBase>);
+        static_assert(!AZStd::totally_ordered_with<Base, const char*>);
+
+        // concept movable
+        static_assert(AZStd::movable<MoveableButNotCopyable>);
+        static_assert(!AZStd::movable<NoMove>);
+
+        // concept movable
+        static_assert(AZStd::copyable<Base>);
+        static_assert(!AZStd::copyable<MoveableButNotCopyable>);
```
Comment (mbalfour-amzn) [inline Code/Framework/AzCore/Tests/AZStd/ConceptsTests.cpp:144]: movable -> copyable

File: Code/Framework/AzCore/Tests/AZStd/TypeTraits.cpp [hunk 5]
```diff
@@ -290,198 +290,317 @@ namespace UnitTest
...
+    constexpr bool NumericCastInvocable = false;
+    template <typename T>
+    constexpr bool NumericCastInvocable<T, AZStd::void_t<decltype(aznumeric_cast<int>(AZStd::declval<T>()))>> = true;
+    TEST(TypeTraits, NumericCastConversionOperatorCompiles)
+    {
+        struct AzNumericCastConvertibleCompileTest
+        {
+            constexpr operator int() { return {}; };
+        };
+        static_assert(NumericCastInvocable<AzNumericCastConvertibleCompileTest>, "aznumeric_cast conversion operator overload is should be compilable");
```
Comment (mbalfour-amzn) [inline Code/Framework/AzCore/Tests/AZStd/TypeTraits.cpp:604]: "is should" -> "should"

**Review by mbalfour-amzn — APPROVED** [review]
LGTM, just a couple of typo nitpicks!

File: Code/Framework/AzCore/AzCore/std/ranges/ranges.h [hunk 6]
```diff
@@ -0,0 +1,1157 @@
...
+
+    template<class T>
+    /*concept*/ constexpr bool viewable_range = range<T> &&
+        ((view<remove_cvref_t<T>> && constructible_from<remove_cvref_t<T>, T>) ||
+            (!view<remove_cvref_t<T>> &&
+                (is_lvalue_reference_v<T> || (movable<remove_reference_t<T>> && !Internal::is_initializer_list<T>))));
+}
+
+// TODO: Move the following code to algorithm.h
+namespace AZStd::ranges
```
[the remark this answers is not in the document]
Comment (lumberyard-employee-dm) [reply Code/Framework/AzCore/AzCore/std/ranges/ranges.h:1078]: There are some circular include dependency issues to resolve if the function was moved.
It is probably better for it to be remain in ranges.h

The issue is that ranges::swap, needs to know about `ranges::swap_range`, but `ranges::swap` is implemented in `concepts.h`, which `algorithm.h` would need to include.

I'll just remove the TODO

File: Code/Framework/AzCore/Tests/AZStd/TypeTraits.cpp (hunk 5 above)
Comment (lumberyard-employee-dm) [reply Code/Framework/AzCore/Tests/AZStd/TypeTraits.cpp:604]: This section of the code didn't change, but I'll fix the typo anyway.

**Review by invertednormal — APPROVED** [review]
Looks good, although I admit to not fully grasping all of the template metaprogramming. Span tests seem a bit light but that can be improved later.

**Review by invertednormal — APPROVED** [review]
Looks great - thanks for the extra span unit tests.

**Pull request merged**